src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1741,25 +1740,61 @@ public override IEnumerator GetEnumerator()
         /// </summary>
         /// <returns></returns>
         public ReadOnlyCollection<NpgsqlDbColumn> GetColumnSchema()
-            => RowDescription == null || RowDescription.Fields.Count == 0
-                ? new List<NpgsqlDbColumn>().AsReadOnly()
-                : new DbColumnSchemaGenerator(_connection, RowDescription, _behavior.HasFlag(CommandBehavior.KeyInfo))
-                    .GetColumnSchema();
+            => GetColumnSchema(async: false).GetAwaiter().GetResult();
 
 #if !NET461
         ReadOnlyCollection<DbColumn> IDbColumnSchemaGenerator.GetColumnSchema()
             => new ReadOnlyCollection<DbColumn>(GetColumnSchema().Select(c => (DbColumn)c).ToList());
 #endif
 
+        /// <summary>
+        /// Asynchronously returns schema information for the columns in the current resultset.
+        /// </summary>
+        /// <returns></returns>
+#if NET // i.e. >= NET5_0
+        public new Task<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
+#else
+        public Task<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
```

Comment:
This introduces GetColumnSchemaAsync even for NET461, which doesn't have the sync version (see above). Better stay consistent on this and not introduce it there.

Reply:
The sync version `GetColumnSchema` is already exposed for NET461, only the `IDbColumnSchemaGenerator.GetColumnSchema` implementation is removed as the interface appeared in NET471.